src/bin/pgcopydb/ld_apply.c

```diff
@@ -991,3 +1008,85 @@ parseSQLAction(const char *query, LogicalMessageMetadata *metadata)
 
 	return true;
 }
+
+
+/*
+ * parseTxnMetadataFile returns the transaction metadata content for the given
+ * trnasanction id
+ */
+bool
+parseTxnMetadataFile(uint32_t xid, LogicalMessageMetadata *metadata, const char *dir)
+{
+	char txnfilename[MAXPGPATH] = { 0 };
+
+	sformat(txnfilename, MAXPGPATH, "%s/%lld.json", dir, (long long) xid);
+
+	/*
+	 * Read the transaction metadata file created by the transform process for
+	 * transactions spanning multiple WAL files. The .json file is generated
+	 * upon encountering the COMMIT statement, but it may take some time to
+	 * become available for transformation. Therefore, we retry indefinitely.
+	 */
+
+	ConnectionRetryPolicy retryPolicy = { 0 };
+
+	int maxT = 1800;            /* 1800s */
```

Comment:
I am not a fan of infinite timeouts. Worst case of a failure here is that the follow supervisor is going to switch to the next apply mode (catchup/replay), right? Let's use a not-too-short timeout there, minutes, not hours. We're waiting for a single transaction to be written to disk, and even if the number of messages per transaction is unbounded, that should happen: after all the logical decoding output plugin did manage to stream the whole transaction to our client code.

Reply:
After timing out here we are returning `false` so this will lead to failed process, follow supervisor also consider it as failed and hence the whole migration fails. I've kept it 15 minutes for now, but it needs more thinking here or probably in follow-up PR as an improvement, what do you think?